oauth/authorization.md

```diff
@@ -15,78 +15,290 @@ Here are links to lessons that should be completed before this lesson:
+	passport.authenticate('auth0', function (err, user, info) {
+		if (err) { return next(err); }
+		if (!user) { return res.redirect('/login'); }
+		req.logIn(user, function (err) {
+		if (err) { return next(err); }
+		const returnTo = req.session.returnTo;
+		delete req.session.returnTo;
+		res.redirect(returnTo || '/user');
+		});
+	})(req, res, next);
+	});
+
+	// Perform session logout and redirect to homepage
+	router.get('/logout', (req, res) => {
+	req.logout();
+	res.redirect('/');
+	});
+
+	module.exports = router;
+
+```
+
+##### Middleware to protect routes
+
+Create a secured middleware to protect routes and ensure they are only accessible if logged in.
+
+If the user is not logged in, the requested route will be stored in the session and the user will be redirected to the login page. Upon successful login, the user will be redirected to the previously requested URL (see callback route above).
+
+``` // lib/middleware/secured.js
+
+	module.exports = function () {
+	return function secured (req, res, next) {
+		if (req.user) { return next(); }
+		req.session.returnTo = req.originalUrl;
+		res.redirect('/login');
+	};
+	};
+
+```
+
+##### Create user profile route
+
+The /user route (the user's profile) should only be accessible if the user is logged in. Use the secured middleware to secure the route.
+
+``` // routes/users.js
+
+	var express = require('express');
+	var secured = require('../lib/middleware/secured');
+	var router = express.Router();
+
+	/* GET user profile. */
+	router.get('/user', secured(), function (req, res, next) {
+	const { _raw, _json, ...userProfile } = req.user;
+	res.render('user', {
+		userProfile: JSON.stringify(userProfile, null, 2),
+		title: 'Profile page'
+	});
+	});
+
+	module.exports = router;
+
+```
+
+##### Create index route
+
+Create an index route to serve the homepage.
+
+``` // routes/index.js
+
+	var express = require('express');
+	var router = express.Router();
+
+	/* GET home page. */
+	router.get('/', function (req, res, next) {
+	res.render('index', { title: 'Auth0 Webapp sample Nodejs' });
+	});
+
+	module.exports = router;
+
+```
+
+##### Making user available in the views
+
+In the views and layouts, it is often necessary to conditionally render content depending on if a user is logged in or not. Other times, the user object might be necessary in order to customize the view.
+
+Create a middleware `lib/middleware/userInViews.js` for this purpose.
+
+``` // userInViews.js
+
+	module.exports = function () {
+	return function (req, res, next) {
+		res.locals.user = req.user;
+		next();
+	};
+	};
+
+```
+
+Include the `lib/middleware/userInViews.js` in the `app.js`
+
+### Independent Practice
+
+echtonica staff will assign pairs. Work together on the Sample App you created during the demonstration to troubleshoot the 'Log in with Google' feature so that when users choose that option, they can successfully log in using their Google credentials.
```

Comment:
Maybe you can give them tips on how to ensure their user persistence is working, and an example of how to include profile data.